dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/SolrOAIReindexer.java

```diff
@@ -0,0 +1,320 @@
+import org.dspace.content.MetadataField;
+import org.dspace.content.MetadataValue;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.ItemService;
+import org.dspace.core.Constants;
+import org.dspace.core.Context;
+import org.dspace.util.SolrUtils;
+import org.dspace.utils.DSpace;
+import org.dspace.xoai.app.BasicConfiguration;
+import org.dspace.xoai.app.XOAIExtensionItemCompilePlugin;
+import org.dspace.xoai.services.api.CollectionsService;
+import org.dspace.xoai.services.api.cache.XOAICacheService;
+import org.dspace.xoai.services.api.cache.XOAIItemCacheService;
+import org.dspace.xoai.services.api.solr.SolrServerResolver;
+import org.dspace.xoai.solr.DSpaceSolrSearch;
+import org.dspace.xoai.solr.exceptions.DSpaceSolrException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+import org.springframework.stereotype.Component;
+
+
+/**
+ * Serves to reindex solr oai core after item has been added, modified or deleted.
+ * Solr document creation (method index(Item item)) taken from XOAI.java.
+ *
+ * @author Marian Berger (marian.berger at dataquest.sk)
+ */
+@Component
+public class SolrOAIReindexer {
+
+    private final XOAICacheService cacheService;
+    private final XOAIItemCacheService itemCacheService;
+
+
+    @Autowired
+    private SolrServerResolver solrServerResolver;
+
+    @Autowired
+    private CollectionsService collectionsService;
+
+    {
+        AnnotationConfigApplicationContext applicationContext =
+                new AnnotationConfigApplicationContext(BasicConfiguration.class);
+        cacheService = applicationContext.getBean(XOAICacheService.class);
+        itemCacheService = applicationContext.getBean(XOAIItemCacheService.class);
+
+    }
+
+    private final Context context = new Context(Context.Mode.READ_ONLY);
+    private final ItemService itemService = ContentServiceFactory.getInstance().getItemService();
+    private final AuthorizeService authorizeService = AuthorizeServiceFactory.getInstance()
+            .getAuthorizeService();
+    private final List<XOAIExtensionItemCompilePlugin> extensionPlugins = new DSpace().getServiceManager()
+            .getServicesByType(XOAIExtensionItemCompilePlugin.class);
+
+    private boolean isPublic(Item item) throws SQLException {
+        // Check if READ access allowed on this Item
+        return authorizeService.authorizeActionBoolean(context, item, Constants.READ);
+    }
+
+
+    private boolean checkIfVisibleInOAI(Item item) throws IOException {
+        SolrQuery params = new SolrQuery("item.id:" + item.getID().toString()).addField("item.public");
+        try {
+            SolrDocumentList documents = DSpaceSolrSearch.query(solrServerResolver.getServer(), params);
+            if (documents.getNumFound() == 1) {
+                return (boolean) documents.get(0).getFieldValue("item.public");
```

Comment:
What if `documents.get(0)` will be null?

Reply:
all taken from XOAI.java